### PR TITLE
Feature/FINP3-4850 Rebase our fork on top of fix8:master

### PR DESCRIPTION
**Pull request by koczwak**

### CODE REVIEW

File: include/fix8/message.hpp [hunk 1]
```diff
@@ -1164,7 +1164,9 @@ class Message : public MessageBase
 	    \return calculated checknum */
 	static unsigned calc_chksum(const char *from, const size_t sz, const unsigned offset=0, const int len=-1)
 	{
-#if defined FIX8_SIZEOF_UNSIGNED_LONG && FIX8_SIZEOF_UNSIGNED_LONG == 8
+// steroid checksum disabled as its bugged for large messages.
+//#if defined FIX8_SIZEOF_UNSIGNED_LONG && FIX8_SIZEOF_UNSIGNED_LONG == 8
+#if 0
```
Comment (fulara) [inline include/fix8/message.hpp]: a little bit of background history here: https://fix8engine.atlassian.net/browse/FX-821
I remember that they have added some commit to fix it, so you might wnat to scan the history whether it has not been fixed yet.

Btw in that jira i have made a test that shows the actual issue.

If you are aligning maybe look for this commit and maybe undo this fix.

To be honest I dont think this will have that much of an impact on the performance, but why not.

File: include/fix8/connection.hpp [hunk 2]
```diff
@@ -355,11 +355,11 @@ class FIXWriter : public AsyncSocket<Message *>
 		quit();
 	}
 
-    bool write_raw(std::function<std::size_t(const Header&)> encode, const char* buffer)
-    {
-        f8_scoped_spin_lock guard(_con_spl);
-        return _session.send_raw_process(encode, buffer);
-    }
+	bool write_raw(const std::function<std::size_t(const Header&)>& encode, const char* buffer)
```
Comment (fulara) [inline include/fix8/connection.hpp:358]: gotta love the fact that this repo is tabbed.

File: include/fix8/thread.hpp [hunk 3]
```diff
@@ -128,8 +130,24 @@ class _f8_threadcore
 #if (FIX8_THREAD_SYSTEM == FIX8_THREAD_PTHREAD)
 		return getid() != get_threadid() ? pthread_join(_tid, nullptr) ? -1 : 0 : -1; // prevent self-join
 #elif (FIX8_THREAD_SYSTEM == FIX8_THREAD_STDTHREAD)
-      if (_thread.get() && _thread->joinable() && getid() != get_threadid())
-			_thread->join();
+		if (_thread.get()) {
+			try {
+				_thread->join();
```
Comment (fulara) [inline include/fix8/thread.hpp]: It would be nice to revisit this fix8 logic and have a proper ownership of the threads.

File: include/fix8/traits.hpp [hunk 4]
```diff
@@ -253,50 +249,32 @@ class presorted_set<unsigned short, FieldTrait, FieldTrait::Compare>
 	  \param sz number of elements in set to copy
 	  \param ftha pointer to field hash array */
 	presorted_set(const_iterator arr_start, const size_t sz, const FieldTrait_Hash_Array *ftha)
-		: _sz(sz), _arr(new FieldTrait[_sz]), _ftha(ftha)
+		: _reserve(), _sz(sz), _arr(new FieldTrait[_sz]), _ftha(ftha)
```
Comment (fulara) [inline include/fix8/traits.hpp]: this commit seems to revert the changes from the lsat few commits, I would consider squashing those.

File: compiler/f8precomp.cpp [hunk 5]
```diff
@@ -159,12 +159,38 @@ int precompfixt(XmlElement& xft, XmlElement& xf, ostream& outf, bool nounique)
 	return 0;
 }
 
+void assertNoDuplicatesWithDifferentNamesButSameId(XmlElement::XmlSet& fldlist) {
+	//ig addition: find out whether there is duplicated name that has different number
```
Comment (fulara) [inline compiler/f8precomp.cpp:163]: this comment seems wrong. too bad there is no example what we wanted to achieve with this function

but i think the different numbers and same name is wrong as well.

**Review by fulara — APPROVED** [review]
Few other remarks.

### conanizing, 
I think it would probably be easier for you if you just ported the conanfiles into this repo, and wrote own cmake for this.
I saw some commit sabout json dependency which ar eno longer here, so i guess you do have some problems in that area.

### 1.4.0
2) Long time ago, when we first started integrating fix8 ( the choice which I will regret forever! ) I have been testing a lot of those scenarios with disconnections and reconnection.

This was using my standard test bench:
1) external party
2) socat
3) our implementation.

Keep turning off and on socat and what I have observed is that sometimes the 1.3.4 was crashing.
Hey - thats when I introduced that nasty exception catch error, just to make it work.

But few months later the 1.4.0 appeared - i was like wow! i can upgrade - they have fixed some stuff in there that was irritating.
So I went into my silly test bench again, and unfortunatlely the 1.4.0 was behaving much worse than the 1.3.4 - there was a deadlock.
I guess i was tired at the time and just dropped the idea of 1.4.0 entirely. 

What we should have really done is either embrace this library altogether, or just never migrate to it. 

### upstream 
I see that you are using latest master, 
there is one PR open that I recommend you to look at whether it has important changes. 
warning: it has some nasty .vcxproj changes in it .. and there is so many of them that its hard to skip it. 

But it may contain some useful changes. 

### That thread shenanigans.
So obviously the implementation done in this codebase around those threads is bad.
I am not sure whether fix that was made minimizes the issue to such extent that we do not care about it, but just few words.

1) There is obviously a wrong thread ownership in this library, checks against self join, and stuff. It would be nice to rework this into proper ownership.
2) I remember that even with this fix, I had come across a weird issue where fix8 started spawning multiple threads like crazy, tried to reproduce but I could not. 

Good luck!

File: include/fix8/message.hpp (hunk 1 above)
Comment (koczwak) [reply include/fix8/message.hpp]: I will reenable it, as it seems to be fixed:
https://godbolt.org/z/jeTbWzYrd
